dlt/destinations/impl/filesystem/layout.py

```diff
@@ -0,0 +1,250 @@
+import re
+from types import TracebackType
+from typing import Any, Dict, List, Optional, Sequence, Set, Type
+
+import pendulum
+from dlt.cli import echo as fmt
+from dlt.common.storages.load_package import ParsedLoadJobFileName
+from dlt.destinations.exceptions import CantExtractTablePrefix, InvalidFilesystemLayout
+from dlt.destinations.impl.filesystem.configuration import (
+    FilesystemDestinationClientConfiguration,
+)
+from typing_extensions import Self
+
+# TODO: ensure layout only has supported placeholders
+SUPPORTED_PLACEHOLDERS = {
+    "schema_name",
+    "table_name",
+    "load_id",
+    "file_id",
+    "ext",
+    "curr_date",
+    "year",
+    "month",
+    "day",
+    "hour",
+    "minute",
+    "dow",
+    "timestamp",
+}
+
+SUPPORTED_TABLE_NAME_PREFIX_PLACEHOLDERS = ("schema_name",)
+
+
+class ExtraParams:
+    def __init__(
+        self,
+        config: FilesystemDestinationClientConfiguration,
+        job_info: Optional[ParsedLoadJobFileName] = None,
+        schema_name: Optional[str] = None,
+        load_id: Optional[str] = None,
+    ) -> None:
+        self.config = config
+        self.job_info = job_info
+        self.load_id = load_id
+        self.schema_name = schema_name
+        self._params: Dict[str, Any] = {}
+
+        self.table_name = None
+        self.file_id = None
+        self.file_format = None
+        if job_info:
+            self.table_name = job_info.table_name
+            self.file_id = job_info.file_id
+            self.file_format = job_info.file_format
+
+    @property
+    def params(self) -> Optional[Dict[str, Any]]:
+        """Process extra params for layout
+        If any value is a callable then we call it with the following arguments
+
+            * schema name,
+            * table name,
+            * load id,
+            * file id,
+            * current datetime
+        """
+        if self.load_id:
+            self._params["load_id"] = self.load_id
+
+        if self.job_info:
+            self._params.update(
+                {
+                    "table_name": self.table_name,
+                    "file_id": self.file_id,
+                    "ext": self.file_format,
+                }
+            )
+
+        # If current_datetime is not given
+        # Then take current moment in time
+        # If current_datetime is callable
+        # Then call it and check it's instance
+        # If the result id DateTime
+        # Then take it
+        # Else exit.
+        if not self.config.current_datetime:
+            self.config.current_datetime = pendulum.now()
```

Comment:
ok this is tricky! we should set a current_datetime only once per instance of factory. so imo you should override `configuration` method and set the `current_datetime` if it is none from a timestamp that is created in factory `__init__`

Reply:
Got it